mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -345,9 +347,179 @@ propagateUnmergeVectorization(T &&dimAndLength,
+    // - their parameters don't match
+    // - skip over singleton dimensions
+    for (size_t j = i; j < unmergeDims.size(); j++) {
+      // Protect against broadcasts
+      if (unmergeParams[j] == 1) {
+        continue;
+      }
+
+      auto it = dimToMerge.find(unmergeDims[j]);
+      if (it == dimToMerge.end() || it->second != mergePair ||
+          paramMatcher[unmergeDims[j]] != unmergeParams[j]) {
+        break;
+      }
+      outputDims.push_back(unmergeDims[j]);
+    }
+
+    // Now we know outputDims contain a group of contiguous dimensions
+    i += std::max(size_t(1), outputDims.size());
+
+    // Update the result with the current group for the mergePair key
+    if (outputDims.size() > 1) {
+      std::sort(outputDims.begin(), outputDims.end(),
+                [&](int a, int b) { return reshuffler[a] < reshuffler[b]; });
+      for (auto d : outputDims) {
+        dimToMerge.erase(d);
+      }
+      result[mergePair].push_back(outputDims);
+    }
+  }
+}
+
+/**
+ * Utility function to filter dimensions that have been spuriously added
+ */
+SmallVector<uint32_t>
+filterDummyDimensions(ArrayRef<uint32_t> dims,
+                      DenseSet<uint32_t> &dummyDimensions) {
+  uint32_t inc = 0;
+  SmallVector<uint32_t> newDims(dims);
+  for (auto &d : newDims) {
+    if (dummyDimensions.contains(d)) {
+      inc++;
+    }
+    d += inc;
+  }
+  return newDims;
+}
+
+/** This is an analysis pass to group the lower dimensions of a Merge
+ * transformation into contiguous groups. E.g., if we have a Merge{8,8,3} [0] ->
+ * [0,2,3] and we know that [2,3] are contiguous in the final representation, we
+ * can split the merge group in
+ * [[0] [2,3]]. This information will be used by the vectorizer. E.g., the
+ * vectorizer can fulfill a vectorization by 4, since 8*3=24 is a multiple of 4.
+ * In other words, every group of dimensions is treated as a single group
+ */
+ContiguousMergesMap findContiguousMerges(ArrayAttr transforms,
+                                         ArrayRef<int64_t> outputShape) {
+  DimToMergeMap dimToMerge;
+  ContiguousMergesMap result;
+  DenseSet<uint32_t> dummyDimensions;
+
+  for (TransformMapAttr transformMap :
+       transforms.getAsRange<TransformMapAttr>()) {
+    for (TransformAttr transform : transformMap.getOps()) {
+      auto transformType = transform.getType();
+      ArrayRef<uint32_t> lowerDims = transform.getLowerDims();
+      ArrayRef<uint32_t> upperDims = transform.getUpperDims();
+      ArrayRef<int64_t> params = transform.getParams();
+
+      switch (transformType) {
+      case TransformType::Merge:
+        for (auto d : lowerDims) {
+          dimToMerge[d] = std::make_pair(transformMap, transform);
+        }
+        break;
+      case TransformType::AddDim:
+        dummyDimensions.insert(upperDims[0]);
+        break;
+      case TransformType::Embed: {
+        SmallVector<int64_t> unmergeParams;
+        for (size_t i = 0; i < params.size() - 1; i++) {
```

Comment:
Note: The parameters on `Embed{}` calls are not guaranteed to be sorted.

Reply:
Second note: you'll need some sort of complexity here to determine where Embed{} acts like Unmerge{}. That is, at our most conservative, we can throw out any Embed{} whose sorted coefficients aren't of the form `1, l_1, l_1 l_2, ...`. If you want to be more clever, we can work out groups of such coefficients ... but maybe that's too much effort for now.